modules/post/windows/manage/ducky.rb

```diff
@@ -0,0 +1,140 @@
+                  '107' => 'ADD', '109' => 'SUBTRACT', '110' => 'DECIMAL', '111' => 'DIVIDE', '112' => 'F1',
+                  '113' => 'F2', '114' => 'F3', '115' => 'F4', '116' => 'F5', '117' => 'F6',
+                  '118' => 'F7', '119' => 'F8', '120' => 'F9', '121' => 'F10', '122' => 'F11',
+                  '123' => 'F12','188' => 'COMMA','190' => 'PERIOD','191' => 'SLASH','48' => '0',
+                  '49' => '1','50' => '2','51' => '3','52' => '4','53' => '5',
+                  '54' => '6', '55' => '7', '56' => '8', '57' => '9'
+                }
+    # Define keypress actions
+    actions = { '0' => 'PRESS', '1' => 'DOWN', '2' => 'UP' }
+    # Parse DuckyScript comments
+    if line.starts_with?('REM ')
+      return
+    end
+    # Cursor movement 
+    # Although this is a bit hacky
+    cursor_line = line.strip
+    if (cursor_line.starts_with?('UP') || cursor_line.starts_with?('DOWN') || cursor_line.starts_with?('LEFT') ||
+      cursor_line.starts_with?('RIGHT') || cursor_line.starts_with?('BACKSPACE') ||
+      cursor_line.starts_with?('TAB') || cursor_line.starts_with?('PAGEUP') || cursor_line.starts_with?('PAGEDOWN') ||
+      cursor_line.starts_with?('HOME') || cursor_line.starts_with?('END') || cursor_line.starts_with?('SPACE') ||
+      cursor_line.starts_with?('RETURN'))
+      session.ui.keyevent_send(key_codes.key(cursor_line).to_i, actions.key('PRESS').to_i)
+    end
+    if line.include?('STRING') || line.include?('STRINGLN') || line.include?('GUI')
+      line_array = line.split(' ', 2)
+      # Write a string and press ENTER
+      if line_array[0] == 'STRINGLN'
+        session.ui.keyboard_send(line_array[1])
+        session.ui.keyevent_send(key_codes.key('RETURN').to_i, actions.key('PRESS').to_i)
+      end
+      # Write a string
+      if line_array[0] == 'STRING'
+        session.ui.keyboard_send(line_array[1])
+      end
+      # Press Windows + R to launch run dialog
+      if line_array[0] == 'GUI'
+        session.ui.keyevent_send(key_codes.key('WINLEFT').to_i, actions.key('DOWN').to_i)
+        session.ui.keyevent_send(key_codes.key('R').to_i, actions.key('PRESS').to_i)
+        session.ui.keyevent_send(key_codes.key('WINLEFT').to_i, actions.key('UP').to_i)
+      end
+    end
+    if line.include?('KEYEVENT')
+      line_array = line.split(' ')
+      # Parse raw keyevents; example format, "KEYEVENT R PRESS"
+      if line_array[0] == 'KEYEVENT'
+        # error handling: ensure that all values are valid before proceeding.
+        if (key_codes.value?(line_array[1].upcase) && actions.value?(line_array[2].upcase))
+          session.ui.keyevent_send(key_codes.key(line_array[1].upcase).to_i, actions.key(line_array[2].upcase).to_i)
+        end
+      end
+    end
+    # Parse F Keys
+    if line.starts_with?('F')
+      session.ui.keyevent_send(key_codes.key(line).to_i, actions.key('PRESS').to_i)
+    end
+  end
+
+  def run
+    return 0 if session.type != 'meterpreter'
```

Comment:
```
fail_with(Failure::None, 'Must have a Meterpreter session to run this module') unless session.type == 'meterpreter'
```

Reply:
I definitely reviewed other modules and found similar uses of `return 0 if session.type != 'meterpreter'` as a meterpreter check before bailing out. Is there are preferred way of doing this?

Reply:
The suggested approach is better; We should probably update the other modules to align :+1:

Reply:
I have updated code in the latest commit that addresses this conversation. I am resolving.